freyja/read_analysis_tools.py

```diff
@@ -388,6 +400,360 @@ def filter(query_mutations, input_bam, min_site, max_site, output, refname):
+                        f'{aa_locus+del_length-1})'
+                    )
+                else:
+                    aa_mut = f'{deletion}({gene}:DEL{aa_locus})'
+                muts_final.append(aa_mut)
+
+            for snp in snps_found:
+
+                # Translate nucleotide muts to amino acid muts
+                locus = int(snp[1:-1])
+
+                gene_info = get_gene(locus)
+                if gene_info is None:
+                    break
+
+                gene, start_site = gene_info
+                codon_position = (locus - start_site) % 3
+                aa_locus = ((locus - codon_position - start_site) // 3) + 1
+
+                ref_codon = ref_genome[locus - codon_position - 1:
+                                       locus - codon_position + 2]
+                ref_aa = ref_codon.translate()
+
+                # Adjust for indels
+                ins_offset = 0
+                if 'I' in x.cigarstring:
+                    for r in ins_offsets:
+                        if locus in range(r[0], r[1]) or locus == r[1]:
+                            ins_offset = ins_offsets[r]
+                del_offset = 0
+                if 'D' in x.cigarstring:
+                    for r in del_offsets:
+                        if locus in range(r[0], r[1]) or locus == r[1]:
+                            del_offset = del_offsets[r]
+
+                read_start = (locus - start) - codon_position + \
+                    ins_offset - del_offset - 1
+                read_end = (locus - start) - codon_position + \
+                    ins_offset - del_offset + 2
+
+                alt_codon = MutableSeq(seq[read_start:read_end])
+
+                if len(alt_codon) % 3 != 0 or len(alt_codon) == 0:
+                    continue  # Possible fail case: codon spans multiple reads
+                alt_aa = alt_codon.translate()
+                if alt_aa == 'X':
+                    continue
+                aa_mut = f'{snp}({gene}:{ref_aa}{aa_locus}{alt_aa})'
+
+                muts_final.append(aa_mut)
+            if coverage_start is None or start < coverage_start:
+                coverage_start = start
+            if coverage_end is None or start + len(seq) > coverage_end:
+                coverage_end = start + len(seq)
+
+        if spans_region:
+            if coverage_start > min_site or coverage_end < max_site:
+                continue
+
+        muts_final = sorted(list(set(muts_final)), key=nt_position)
+        name = ' '.join([str(mut).replace(' ', '') for mut in muts_final])
+        if len(name) > 1:
+            if name not in co_muts:
+                co_muts[name] = 1
+            else:
+                co_muts[name] += 1
+            coverage[name] = (coverage_start, coverage_end)
+    samfile.close()
+
+    df = pd.DataFrame()
+    df['Covariants'] = [k for k in co_muts]
+    df['Count'] = [co_muts[k] for k in co_muts]
+    df['Frequency'] = df['Count'] / all_reads_count
```

Comment:
Dividing out by the "all reads" count is a bit misleading. For example, for the test.bam file in data, if i look at the range of 22900 to 23200, i get a super small fraction with combined T22917G(S:L452R) and C22995A(S:T478K) but it's probably closer to 20% of reads that span that whole region. This will get even more weird as we look at fragmented reads, so I think we might want to just drop frequency as a readout altogether and just stick with counts.